Automatically bind CPU OMP Threads of a rank to CPU ids of a NUMA node.

louie-tsai opened the pull request:
**[Current Implementation]**
In order to have good performance for Tensor Parallel or Pipeline Parallel on CPU, users need to do proper CPU OMP Threads binding to avoid perf degradation causing by multiple threads on same CPU core.

Here is the current run instructions for CPU OMP THREADS BIND.
`OMP_NUM_THREADS=32 VLLM_CPU_OMP_THREADS_BIND="0-31|32-63|64-95|96-127"  python3 -m vllm.entrypoints.openai.api_server --model meta-llama/Meta-Llama-3.1-8B-Instruct   --device cpu  -tp=4 --distributed-executor-backend mp `

**[Problem Statement]**
However, CPU ids might change among OSes, and different CPU SKU also have different CPU numbers.
Moreover, users might need to check cpus_allow_list also to only bind to allowed CPU cores.
It requires users to check their environment first, and set the Binding properly.
In some cases like cluster deployment using Kubernetes, users won't know the CPU ids before the deployments, so it is hard to set the deployment scripts like k8s yaml file correctly.

**[Proposed Solution]**
Therefore, we introduce a new feature to automatically bind CPU OMP threads of a rank to CPU ids of a allowed NUMA node according to cpus_allowed_list.
Therefore, no need to set the VLLM_CPU_OMP_THREADS_BIND according to users env, and CPU worker will set it automatically for users.
New run instructions will like below one, and it is also easier to have Tensor Parallel support for k8s deployment environment.
` python3 -m vllm.entrypoints.openai.api_server --model meta-llama/Meta-Llama-3.1-8B-Instruct   --device cpu  -tp=4 --distributed-executor-backend mp `

Overall, it will bind a rank/world to a allowed numa node, so 4 numa nodes will be used for tp=4 pp=1.
2 numa nodes will be used for tp=2 pp=1.
if current environment only allow 2 numa nodes, it will return errors.

**[Related Environment variables]**

VLLM_CPU_OMP_THREADS_BIND  : By setting to auto, the OpenMP threads of each rank are bounds to the CPU cores in each NUMA node. Default value is auto.
VLLM_CPU_NUM_OF_RESERVED_CPU:  specify the number of CPU cores which are not dedicated to the OpenMP threads for each rank. The variable only takes effect when VLLM_CPU_OMP_THREADS_BIND is set to auto. Default value is 0.


**[More Details]**
you could see in the below diagram.
Even vllm server doesn't get input for VLLM_CPU_OMP_THREADS_BIND, and set omp_cpuids to all.
CPU worker will automatically overwrite the local_omp_cpuid according to current system numa configuration.
![image](https://github.com/user-attachments/assets/04de5782-d101-4f52-a9db-6771aa89eaee)

<!--- pyml disable-next-line no-emphasis-as-heading -->

louie-tsai committed:
automatically bind OMP Threads of a rank to CPU ids of a NUMA node.

Signed-off-by: Tsai, Louie <louie.tsai@intel.com>

louie-tsai committed:
handle hyperthreading on case

Signed-off-by: Tsai, Louie <louie.tsai@intel.com>

louie-tsai committed:
only use cpu cores in the allowed list

Signed-off-by: Tsai, Louie <louie.tsai@intel.com>

bigPYJ1151 commented on vllm/worker/cpu_worker.py:
```diff
@@ -155,11 +155,51 @@ def __init__(
+                    logger.info("local_omp_cpuid: %s",
+                        self.local_omp_cpuid)
+            else:
+                logger.info("[ERROR] NO AUTO OMP Bind support. Please install py-libnuma to have the feature")
```
```suggestion
                logger.warning("Auto thread-binding is not supported due to the lack of package numa and psutil, fallback to no thread-binding. To get better performance, please try to manually bind threads.")
```

bigPYJ1151 reviewed (commented):
@louie-tsai  Thanks for your good work and apology for the late response. I have some suggestions, please take a look. 

BTW, can you help to update the CPU document about the PR as well? You can find it at ```docs/source/getting_started/installation/cpu.md```, thanks :)

askervin commented on vllm/worker/cpu_worker.py:
```diff
@@ -155,11 +155,51 @@ def __init__(
+                    logger.info("[ERROR] NO AUTO OMP Bind support because request world size: %d is more than allowed numa_size: %d",
+                            world_size, len(node_to_cpus))
+                else:
+                    rank_to_cpus=str(node_to_cpus[self.rank][0]) + '-' + str(node_to_cpus[self.rank][cpu_count_per_numa - 1 - num_of_no_bind_cpu])
```
Building the `rank_to_cpus ` string assumes that CPU ids in `node_to_cpus[self.rank]` are contiguous. But it could be that there are CPUs 0-39,240-279. If so, this means that `rank_to_cpus` would contains CPUs 0-279. The safest option would be to construct the CPU list without assuming anything about how CPUs are numbered inside a NUMA node. (There's a lot variation in different platforms.)

louie-tsai replied:
@askervin 
We used `info.node_to_cpus(i)` to get the list of CPU ids in a numa node, so we don't assume the CPU id will be continues in a numa node.  

For CPUs 0-39,240-279, we only get node_to_cpus = [0-39]  since 240-279 is also in numa node 0 when HT is on.

askervin commented on vllm/worker/cpu_worker.py:
```diff
@@ -155,11 +155,51 @@ def __init__(
+                cpu_count = psutil.cpu_count(logical=False)
+                cpus_allow_list = psutil.Process().cpu_affinity()
+                numa_size = info.get_num_configured_nodes()
+                cpu_count_per_numa = cpu_count // numa_size
```
Some systems include NUMA nodes with CPUs and other NUMA nodes without CPUs. CPU-less NUMA nodes might be, for instance HBM (high-bandwidth memory), PMEM (persistent memory) or CXL memory blocks.

I'd suggest calculating `cpu_count_per_numa` by dividing the number of CPUs with the number of NUMA nodes that contain CPUs.

Or, to be even more accurate, calculate first `node_to_cpus` so that 

`node_to_cpus[node] = intersection of cpus on the node and the set of allowed CPUs`

...and skip all nodes where the intersection is empty.

After this, if all node_to_cpus contain the same number of CPUs, there is your cpu_count_per_numa. And then your algorithm will work even if allowed CPUs would be equal-sized chunks from separate NUMA nodes. I think this would be very nice.

If CPU sets in nodes_to_cpus are of different size, it would be fine to print a warning about not doing auto affinity to ranks.

louie-tsai replied:
@askervin 
thanks for your detailed input.  cpu_count_per_numa is only used to get the index of last CPU in one node.  it doesn't use to find out node_to_cpus 
The main logic to find CPU ids for each node are in below codes, and it should address your request `"node_to_cpus[node] = intersection of cpus on the node and the set of allowed CPUs` 

![image](https://github.com/user-attachments/assets/d4d219ab-f7ea-49e6-81bf-a3f431af68d8)

also the info.node_to_cpus(i) should cover cpu id only in one numa node. In general, we should have same number of cpus per node. (not the case for GNR SNC=3).
I assume that info.node_to_cpus(i) will return CPU-less NUMA nodes. Please correct me if I have misunderstandings.

askervin replied:
Let's use a two-socket Xeon MAX CPU as an example. It has high-bandwidth memory (HBM) integrated directly into CPU package, and DDR5 memory DIMMs next to each socket.

Assume that this system is configured to expose 4 NUMA nodes. (Depending on configuration, it could have even 16 nodes in real-life.) Assume that nodes 0 and 1 include both CPUs and memory (DDR5), while NUMA nodes 2 and 3 include only memory (HBM) and no CPUs.

`info.get_numa_configured_nodes()` returns the number of all NUMA nodes, that is 4. This would be the `numa_size`. Let's assume that

`info.nodes_to_cpus(0)` returns [0,1,..., 30, 31, 64, 65, ..., 94, 95]

`info.nodes_to_cpus(1)` returns []

`info.nodes_to_cpus(2)` returns [32, 33, ..., 62, 63, 96, 97, ..., 126, 127]

`info.nodes_to_cpus(3)` returns []

This gives you `nodes_to_cpus = [[0-31, 64-95], [], [32-63, 96-127], []]`. And `cpu_count_per_numa = cpu_count / numa_size = 64 / 4 = 16`

Now, in the case of rank=0,

`rank_to_cpus=str(node_to_cpus[self.rank][0]) + '-' + str(node_to_cpus[self.rank][cpu_count_per_numa - 1 - num_of_reserved_cpu])`

gives "0-16". rank=1 crashes on IndexError.

And this example is still a kind of easy, CPU ids of thread0 and thread1 in the NUMA nodes are continuous. But there are platforms where they are not. The code should work even if
`info.nodes_to_cpus(0)` returns [0,2,4,6]
`info.nodes_to_cpus(1)` returns [1,3,5,7]

NUMA ordering and CPU numbering can be quite exciting, so let's make any assumptions on it.


So what I'm suggesting is:
- Do not require `set(info.node_to_cpus(i)).issubset(cpus_allow_list)`. This requirement is very hard, because it means that the resource policy that is managing the server has been able to allocate every single CPU from a NUMA node. But there are many other containers (other vLLM containers, databases, whatever servers) that might have exclusive CPUs from every node, and therefore cpus_allow_list may not include all CPUs from any NUMA node. But it if includes 16 CPUs from one NUMA node and 16 CPUs from another NUMA node, this automatic optimization could still work.

- So what do you think, if instead of this requirement, you would construct node_to_cpus like `node_to_cpus.append(set(info.node_to_cpus(i)).intersection(cpus_allow_list))` whenever the intersection is not empty? This would automatically solve the problem of CPU-less NUMA node (the IndexError that I mentioned above). And it would enable running nicely in GNR SNC3, if only the resource policy allocates equal number of CPUs from every node.

louie-tsai replied:
moving the implementations into a separate function accordingly.  also handled the non-continuous cpd-id cases.

louie-tsai committed:
address Li, Jiang's feedback

Signed-off-by: Tsai, Louie <louie.tsai@intel.com>

louie-tsai committed:
Apply suggestions from code review

Co-authored-by: Li, Jiang <bigpyj64@gmail.com>
Signed-off-by: Tsai, Louie <louie.tsai@intel.com>

louie-tsai commented on vllm/envs.py:
```diff
@@ -374,6 +376,16 @@ def maybe_convert_int(value: Optional[str]) -> Optional[int]:
 
+    # (CPU backend only) CPU core ids auto bound by OpenMP threads.
+    # OpenMP threads of a rank will auto bound to a NUMA node.
+    "VLLM_CPU_AUTO_OMP_THREAD_NUMA_BIND":
```
agreed. changed it accordingly.

louie-tsai commented on requirements/cpu.txt:
```diff
@@ -19,3 +19,5 @@ datasets # for benchmark scripts
 
 # cpu cannot use triton 3.3.0
 triton==3.2.0; platform_machine == "x86_64"
+py-libnuma # for auto omp thread numa bind
```
Thanks. changed it accordingly.

louie-tsai commented on vllm/worker/cpu_worker.py:
```diff
@@ -155,11 +155,51 @@ def __init__(
+                            world_size, len(node_to_cpus))
+                else:
+                    rank_to_cpus=str(node_to_cpus[self.rank][0]) + '-' + str(node_to_cpus[self.rank][cpu_count_per_numa - 1 - num_of_no_bind_cpu])
+                    logger.info("omp_cpuids: %s, rank: %d, world_size:%d, tp_size: %d, pp_size: %d",
```
agreed. changed it accordingly.

louie-tsai committed:
change default to auto

Signed-off-by: Tsai, Louie <louie.tsai@intel.com>

bigPYJ1151 reviewed (commented):
We are closing to be ready.

Please add some note to the CPU doc for the new added env. And fix the code-style checks, you can refer to [this](https://docs.vllm.ai/en/latest/contributing/overview.html#developing) for auto format and lint.

louie-tsai committed:
Update vllm/worker/cpu_worker.py

Co-authored-by: Li, Jiang <bigpyj64@gmail.com>
Signed-off-by: Tsai, Louie <louie.tsai@intel.com>

louie-tsai committed:
Update vllm/worker/cpu_worker.py

Co-authored-by: Li, Jiang <bigpyj64@gmail.com>
Signed-off-by: Tsai, Louie <louie.tsai@intel.com>

louie-tsai committed:
change from issubset to intersect according to Antti's feedback

Signed-off-by: Tsai, Louie <louie.tsai@intel.com>

louie-tsai committed:
pre-commit fix

Signed-off-by: Tsai, Louie <louie.tsai@intel.com>

louie-tsai committed:
get cpus_allow_list in main process
the forked process of each rank might get empty cpus_allow_list instead of right cpus list

Signed-off-by: Tsai, Louie <louie.tsai@intel.com>

louie-tsai committed:
Update cpu.md

Signed-off-by: Tsai, Louie <louie.tsai@intel.com>

louie-tsai commented:
> We are closing to be ready.
> 
> Please add some note to the CPU doc for the new added env. And fix the code-style checks, you can refer to [this](https://docs.vllm.ai/en/latest/contributing/overview.html#developing) for auto format and lint.

@bigPYJ1151  updated the cpu.md accordingly

louie-tsai committed:
set default value to all to simplify if else statement

Signed-off-by: Tsai, Louie <louie.tsai@intel.com>

louie-tsai committed:
pre-commit fix

Signed-off-by: Tsai, Louie <louie.tsai@intel.com>

louie-tsai commented:
@bigPYJ1151  let me know any change needed. all tests are passed, so hopefully it could be approved soon.

louie-tsai committed:
fix correctness test for gpu

Signed-off-by: Tsai, Louie <louie.tsai@intel.com>

bigPYJ1151 commented on vllm/executor/mp_distributed_executor.py:
```diff
@@ -108,6 +108,10 @@ def _init_executor(self) -> None:
         distributed_init_method = get_distributed_init_method(
             get_ip(), get_open_port())
+        # Get cpus_allow_list for main process and its ranks
+        import psutil
+        cpus_allow_list = psutil.Process().cpu_affinity()
```
Why not getting ```cpus_allow_list``` in the ```CPUWorker```? I think worker children processes have same CPU affinity with the parent before the OMP thread binding.

louie-tsai replied:
@bigPYJ1151  
I tested it on my EMR env. only parent process got right cpus_allow_list. children processes got 0 for cpus_allow_list and failed.  I highlight related logs in red below.
![image](https://github.com/user-attachments/assets/6c33f151-ab91-471d-a9d8-1b35ed51556e)

Therefore, I moved the codes to executor under parent process.

bigPYJ1151 replied:
I think it is due to the fork method.
Can you please help to try again with ```VLLM_WORKER_MULTIPROC_METHOD=spawn```?
If it works, you can add it to the CPU platform settings:
https://github.com/vllm-project/vllm/blob/5a8641638a5ec6cf089490d6b7f474b135ff590e/vllm/platforms/cpu.py#L147-L168

louie-tsai replied:
thanks. it works. change default value from fork to spawn in envs.py

louie-tsai replied:
addressed it accordingly

louie-tsai committed:
fix forked processes of ranks issue for cpus_allow_list by changing to spawn

Signed-off-by: Tsai, Louie <louie.tsai@intel.com>

louie-tsai committed:
change to handle non-continuous cpu ids case

Signed-off-by: Tsai, Louie <louie.tsai@intel.com>

louie-tsai committed:
move auto binding into a function

Signed-off-by: Tsai, Louie <louie.tsai@intel.com>

louie-tsai committed:
Update cpu.md to address Antti's feedback

Signed-off-by: Tsai, Louie <louie.tsai@intel.com>

bigPYJ1151 commented:
Hi @louie-tsai Thanks for your refinement. This PR is in a good shape I think. 

Just now the CPU V1 support was merged into the main. Would you please help to copy the auto binding procedure to [V1 worker](https://github.com/vllm-project/vllm/blob/main/vllm/v1/worker/cpu_worker.py)?

After that we can try to merge this PR, thanks :)

louie-tsai commented:
> Hi @louie-tsai Thanks for your refinement. This PR is in a good shape I think.
> 
> Just now the CPU V1 support was merged into the main. Would you please help to copy the auto binding procedure to [V1 worker](https://github.com/vllm-project/vllm/blob/main/vllm/v1/worker/cpu_worker.py)?
> 
> After that we can try to merge this PR, thanks :)

Wonderful to have V1 merged. will do.

louie-tsai committed:
add auto binding into V1 feature

Signed-off-by: Tsai, Louie <louie.tsai@intel.com>

Isotr0py reviewed (approved):
```
(VllmWorker rank=0 pid=1186204) INFO 06-05 16:59:15 [cpu_worker.py:145] auto thread-binding list: 0,1,2,3,4,5,6,7,8,9,10,11
(VllmWorker rank=1 pid=1186205) INFO 06-05 16:59:15 [cpu_worker.py:145] auto thread-binding list: 12,13,14,15,16,17,18,19,20,21,22,23
(VllmWorker rank=1 pid=1186205) INFO 06-05 16:59:15 [cpu_worker.py:51] OMP threads binding of Process 1186205:
(VllmWorker rank=1 pid=1186205) INFO 06-05 16:59:15 [cpu_worker.py:51]  OMP tid: 1186205, core 12
(VllmWorker rank=1 pid=1186205) INFO 06-05 16:59:15 [cpu_worker.py:51]  OMP tid: 1186674, core 13
(VllmWorker rank=1 pid=1186205) INFO 06-05 16:59:15 [cpu_worker.py:51]  OMP tid: 1186675, core 14
(VllmWorker rank=1 pid=1186205) INFO 06-05 16:59:15 [cpu_worker.py:51]  OMP tid: 1186676, core 15
(VllmWorker rank=1 pid=1186205) INFO 06-05 16:59:15 [cpu_worker.py:51]  OMP tid: 1186677, core 16
(VllmWorker rank=1 pid=1186205) INFO 06-05 16:59:15 [cpu_worker.py:51]  OMP tid: 1186678, core 17
(VllmWorker rank=1 pid=1186205) INFO 06-05 16:59:15 [cpu_worker.py:51]  OMP tid: 1186679, core 18
(VllmWorker rank=1 pid=1186205) INFO 06-05 16:59:15 [cpu_worker.py:51]  OMP tid: 1186680, core 19
(VllmWorker rank=1 pid=1186205) INFO 06-05 16:59:15 [cpu_worker.py:51]  OMP tid: 1186681, core 20
(VllmWorker rank=1 pid=1186205) INFO 06-05 16:59:15 [cpu_worker.py:51]  OMP tid: 1186682, core 21
(VllmWorker rank=1 pid=1186205) INFO 06-05 16:59:15 [cpu_worker.py:51]  OMP tid: 1186683, core 22
(VllmWorker rank=1 pid=1186205) INFO 06-05 16:59:15 [cpu_worker.py:51]  OMP tid: 1186684, core 23
(VllmWorker rank=1 pid=1186205) INFO 06-05 16:59:15 [cpu_worker.py:51] 
(VllmWorker rank=0 pid=1186204) INFO 06-05 16:59:16 [cpu_worker.py:51] OMP threads binding of Process 1186204:
(VllmWorker rank=0 pid=1186204) INFO 06-05 16:59:16 [cpu_worker.py:51]  OMP tid: 1186204, core 0
(VllmWorker rank=0 pid=1186204) INFO 06-05 16:59:16 [cpu_worker.py:51]  OMP tid: 1186720, core 1
(VllmWorker rank=0 pid=1186204) INFO 06-05 16:59:16 [cpu_worker.py:51]  OMP tid: 1186721, core 2
(VllmWorker rank=0 pid=1186204) INFO 06-05 16:59:16 [cpu_worker.py:51]  OMP tid: 1186722, core 3
(VllmWorker rank=0 pid=1186204) INFO 06-05 16:59:16 [cpu_worker.py:51]  OMP tid: 1186724, core 4
(VllmWorker rank=0 pid=1186204) INFO 06-05 16:59:16 [cpu_worker.py:51]  OMP tid: 1186725, core 5
(VllmWorker rank=0 pid=1186204) INFO 06-05 16:59:16 [cpu_worker.py:51]  OMP tid: 1186726, core 6
(VllmWorker rank=0 pid=1186204) INFO 06-05 16:59:16 [cpu_worker.py:51]  OMP tid: 1186727, core 7
(VllmWorker rank=0 pid=1186204) INFO 06-05 16:59:16 [cpu_worker.py:51]  OMP tid: 1186728, core 8
(VllmWorker rank=0 pid=1186204) INFO 06-05 16:59:16 [cpu_worker.py:51]  OMP tid: 1186729, core 9
(VllmWorker rank=0 pid=1186204) INFO 06-05 16:59:16 [cpu_worker.py:51]  OMP tid: 1186730, core 10
(VllmWorker rank=0 pid=1186204) INFO 06-05 16:59:16 [cpu_worker.py:51]  OMP tid: 1186731, core 11
(VllmWorker rank=0 pid=1186204) INFO 06-05 16:59:16 [cpu_worker.py:51]
```

Given the auto bind results locally, this PR LGTM! Thanks for this improvement!

louie-tsai commented on vllm/v1/worker/cpu_worker.py:
```diff
@@ -99,3 +103,50 @@ def execute_model(
+        rank_to_cpus = self.local_omp_cpuid
+        # Setup OpenMP thread affinity based on NUMA nodes automatically
+        world_size = self.vllm_config.parallel_config.world_size
+        from importlib import util
```
do we want to remove the numa and psutil checking assuming since it is installed via cpu.txt?

louie-tsai replied:
thanks for the input. addressed it accordingly.

xuechendi commented:
Hi @DarkLight1337 @bigPYJ1151, if the PR looks good to you, please help to approve.

louie-tsai committed:
address isotr0py's feedback

Signed-off-by: Tsai, Louie <louie.tsai@intel.com>

jikunshang commented on vllm/platforms/cpu.py:
```diff
@@ -208,6 +208,9 @@ def check_and_update_config(cls, vllm_config: VllmConfig) -> None:
         os.environ["TORCHINDUCTOR_COMPILE_THREADS"] = "1"
 
+        # Share the cpusets list among ranks by spawning process instead
+        os.environ["VLLM_WORKER_MULTIPROC_METHOD"] = "spawn"
```
I think we double set this. duplicate to  L199.  may due to rebase not auto merge this change.
